### DISCUSSION TITLE
Which sensor data is used to create the /odom data for the turtlebot 4?

### DISCUSSION THREAD

**Discussion by WaitingForAll** [turtlebot/turtlebot4#306] [Q&A]
I havn't found any information in the turtlebot4 packages. I someone could point me to the right direction, this would be really great!

**Comment by bmaxdk** [comment]
In this [Turtlebot4 link here](https://clearpathrobotics.com/turtlebot-4/#:~:text=Onboard%20sensors%20include%20the%20OAK,connected%20ROS%202%20API), the sensors packed addressed that **Onboard sensors include the OAK-D spatial AI stereo camera, 2D Lidar, IMU, optical floor tracking sensor, wheel encoders, infrared, cliff, bump, slip detection and more. All onboard sensors are available as ROS topics via the network-connected ROS 2 API**

`IMU` and `wheel encoders` are the  primary sensors used for odometry data in many robotic platforms, including the TurtleBot 4, given their direct role in tracking movement and orientation. The other sensors mentioned in the link, the optical floor tracking sensor, 2D Lidar, and others, assist in overall navigation and environment mapping but are not the primary source for /odom data.

You may also want to check Joint State Publisher. This node is used primarily for publishing the states of joints (like angles, velocities, and efforts) in a robot.

**Reply by slowrunner** [reply]
Does the optical flow sensor data feed into the /odom topic?  

You mention "not the primary source", but is it fused into the initial /odom topic?

Is it only fused later with "encoder+IMU" /odom in the SLAM/Nav2 packages into /pose topics?
